js-packages/@prestojs/rest/src/Endpoint.ts

```diff
@@ -509,54 +643,107 @@ export default class Endpoint<ReturnT = any> {
+                            // is not paginated.
+                            const paginationState = cls.defaultConfig.getPaginationState(
+                                paginator,
+                                returnVal
                             );
+                            if (paginationState) {
+                                paginator.setResponse(paginationState);
+                                data = paginationState.results;
+                            } else {
+                                // TODO: If you specify a paginator and the response is not paginated should
+                                // we warn? I think yes as I can't think of a reason why not but may need to
+                                // revisit this if we find a usecase
+                                console.warn(
+                                    'A paginator was defined but the response was not paginated. Paginator state has not been updated.'
+                                );
+                            }
                         }
+                        return this.transformResponseBody ? this.transformResponseBody(data) : data;
+                    });
+            const executeWithMiddleware = (): Promise<ReturnT> => {
+                const middleware = [...this.middleware];
+                let lastMiddleware;
+                const next = async (
+                    url: string,
+                    requestInit: EndpointRequestInit
+                ): Promise<ReturnT> => {
+                    let errors: string[] = [];
+                    if (typeof url !== 'string') {
+                        errors.push(`'url' arg from middleware was not a string, received: ${url}`);
                     }
-                    return this.transformResponseBody ? this.transformResponseBody(data) : data;
-                });
+                    if (!requestInit || typeof requestInit != 'object') {
+                        errors.push(
+                            `'requestInit' arg from middleware was not an object, received: ${requestInit}`
+                        );
+                    }
+                    if (errors.length > 0) {
+                        const err = errors.join('\n');
+                        throw new Error(
+                            `Bad middleware implementation; invalid arguments\n\n${err}\n\nOccurred in middleware:\n\n${lastMiddleware}`
+                        );
+                    }
+                    const nextMiddleware = middleware.shift();
+                    if (!nextMiddleware) {
+                        return runFetch(url, requestInit);
+                    }
+                    lastMiddleware = nextMiddleware;
+                    const r = nextMiddleware(url, requestInit, next, {
+                        execute: executeWithMiddleware,
+                        executeOptions: options,
+                    });
```

Comment:
- If there is some middleware that eg attempts to replay a request in the event of a 403 the then won't the first run have already removed everything from the `middleware` list, so the attempt at replaying the query will have no subsequent middleware?
- In the same example above, if middleware modifies the requestInit object then that will persist through both replays (eg if it adds a query param then the query param will have been added twice on the second replay)

It seems to me that before each invocation of a middleware next() needs to clone everything that's about to be passed into that middleware in case it's modified

Reply:
* no, at start of function we have `const middleware = [...this.middleware];` - it's copied. replaying it replays again with all middleware
* correct; as mentioned in your comment earlier mutations would change the context too so changing `next(url, requestInit);` to `next(url, deepClone(requestInit));` will resolve that to

Reply:
I think there's a misunderstanding - it does clone it (see where it kicks off it calls mergeRequestInit)

It's not expected that middleware would clone it - it's expected they mutate it but mutations only apply to current middleware execution chain.

Reply:
Can you mention this in the header for `execute()`.

I'm still unclear on how middleware would go about replaying a request though.

Consider the 403 example above.

I had thought that as the request came back up the middleware stack, the 403 interceptor would keep the original request unresolved, would trigger a login popup, log in again, replay the request from where we were at, and then return the request from that point on. If the middleware has already been removed then how can you replay the request from this point on?

What I thought:

- execute()
  - executeWithMiddleware()
    - next()
      - customMiddlewareA
        - next()
          - 403Handler()
            - next()
              - customMiddlewareB() [modifies the request]
                - next()
                  - actual fetch request, fails with 403
                - resolve Promise
            - resolve Promise
            - 403 handler does not resolve promise, instead triggers a completely separate login cycle
            - 403 handler now tries to call next() again to replay the request from this point
              - but there's nothing left in the middleware array anymore
              - even if there was, customMiddlewareB will be called a second time and so will modify the request again. That's ok if it's idempotent but it may not be

Reply:
Edit: reading in order helps.

With MiddlewareContext it would look like:

- execute()
  - executeWithMiddleware()
    - next()
      - customMiddlewareA
        - next()
          - 403Handler()
            - next()
              - customMiddlewareB() [modifies the request]
                - next()
                  - actual fetch request, fails with 403
                - resolve Promise with return from next()
            - 403 handler sees a 403 so does _not_ resolve promise, instead triggers a completely separate login cycle
            - 403 handler now calls MiddlewareContext.execute() again to trigger the request from the start

Is that correct?

Reply:
Not quite - the original promise wouldn't resolve (at least not how I would do it). The 403Handler instead returns a new promise that only resolves once the login happens. When it resolves it resolves to the promise returned from the replayed request after the successful login.